src/woollypully/DESIGN.md

```diff
@@ -0,0 +1,69 @@
+# woollypully
+
+Woollypully uses Sweater Comb rules for API governance over Kubernetes deployments.
+
+## Why?
+
+Prevent releases that do not conform to API standards from seeing the light of day.
+
+## Just-In-Time Gating
+
+### API Discovery
+
+Using a set of well-known endpoints on each microservice, we can:
+
+- Discover what APIs are provided (a service may provide multiple)
+- Match on those we need to govern (public OpenAPI ones)
+- Apply checks to all the OpenAPI versions provided by those
+
+The main entrypoint in this woollypully sidecar makes use of these endpoints to coordinate the gating.
+
+### Comparing current and proposed
+
+TODO: open problem...
+
+The API discovery isn't exposed outside the pod, so how would we compare with the current running service?
+
+### Sidecar containers and readiness probes
+
+A WoollyPully sidecar container may be injected into pods. The sidecar discovers the APIs, checks their OpenAPI contracts for compliance, and makes a pass / fail determination.
```

Comment:
I'm wondering here if Cerberus shouldn't be responsible for this job with the help of a `WoollyPully` library.

Rationales:
- Cerberus should already be part of the "service" stack
- Cerberus already parses the API discovery and OpenAPI specification
- Cerberus is already responsible for some part of the API governance

Reply:
> A WoollyPully library might look like a sidecar, since our API linting (this project) is Typescript, and Cerberus is Go. But if Cerberus can orchestrate it, that might be interesting...

Yeah, the orchestration bit is the difficult bit. Maybe there are ways: https://github.com/metacall/golang-typescript-example.

> One property that I'd want here though, that you shouldn't be able to opt out of, or forget to add, or configure WoollyPully. It must be something applied "from the outside" to deployments as policy.

This is also something that we want for Cerberus, being injected instead of being defined by devs.